### PR TITLE
remove materialization from DataFrameConcatOperation

### PR DESCRIPTION
**Pull request by KathiaBarahona**
**Issue**: https://github.com/objectiv/objectiv-analytics/issues/401

### Description
Previously `DataFrameConcatOperation` materialized each dataframe to be concatenated after calling `_fill_missing_series`. The reason of the materialization was that the class will use the result from `_get_series`, which returns the final series of the dataframe. The issue of this was that each expression made referenced to an existent column of the subquery.

In order to remove the materialization, I decided to remove `_fill_missing_series` method and instead of this, construct a new expression for the columns that don't exist in that base_node. This considerably made smaller the query.

### Checklist
- [X] Tests
- [X] Updated docstrings

### CODE REVIEW

**Review by thijs-obj — COMMENTED** [review]
Question: in `SeriesConcatOperation` there is still a call to `materialize()`. Is that still needed?

**Comment by thijs-obj** [comment]
The DataFrame part looks good to me. Question about the Series part

File: bach/bach/concat.py [hunk 1]
```diff
@@ -143,35 +143,19 @@ def _get_overridden_objects(self) -> List[DataFrame]:
             if self.ignore_index:
```
Comment (jansentom) [inline bach/bach/concat.py:143]: `df = obj.copy_override()` can be written as `df = obj.copy()`. That will do exactly the same, but shorter.

File: bach/tests/functional/bach/test_df_append.py [hunk 2]
```diff
@@ -161,3 +161,34 @@ def test_append_w_different_dtypes() -> None:
...
+def test_append_w_non_materialized_df() -> None:
+    caller_pdf = pd.DataFrame({'b': [1, 2, 3, 4], 'c': [1, 2, 3, 4]})
+    other_pdf = pd.DataFrame({'b': [1.1, 2.2, 3.3, 4.4, 5.5], 'c': [5, 6, 7, 8, 9]})
+
+    caller_df = get_from_df('caller_df', caller_pdf)
+    other_df = get_from_df('other_df', other_pdf)
+
+    caller_df['d'] = caller_df['b'] + caller_df['c']
+
+    caller_df = caller_df.reset_index(drop=True)
```
Comment (jansentom) [inline bach/tests/functional/bach/test_df_append.py:175]: `reset_index` is not required here. `set_index()` below already discards the old index.

**Review by jansentom — APPROVED** [review]
Two small optimizations, but LGTM in general. Very nice cleanup of the generated SQL!

**Pull request merged**